### PR TITLE
fix bad rewritten result for foreign key SQL

### LINKED ISSUE #10171: Bad rewritten result for foreign key SQL
### DB
PostgreSQL

### SQL
```sql
alter table bmsql_district add constraint d_warehouse_fkey foreign key (d_w_id) references bmsql_warehouse (w_id);
```

### Intro
- `bmsql_district` and `bmsql_warehouse` are sharding tables but not binding tables.

### Actual rewritten result
alter table bmsql_district_0 add constraint d_warehouse_fkey foreign key (d_w_id) references bmsql_warehouse (w_id);

### Expected rewritten result
alter table bmsql_district_0 add constraint d_warehouse_fkey foreign key (d_w_id) references bmsql_warehouse_0 (w_id);

### PR DESCRIPTION
**Pull request by strongduanmu**
Fixes #10171.

Changes proposed in this pull request:
- fix bad rewritten result for foreign key SQL when user config single data node sharding config
- refactor ShardingTableBroadcastRoutingEngine to support different table combination
- add some test cases for table rewrite with different table combination

### CODE REVIEW

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-route/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingAlterTableStatementValidator.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+ * Sharding alter table statement validator.
+ */
+public final class ShardingAlterTableStatementValidator extends ShardingDDLStatementValidator<AlterTableStatement> {
+    
+    @Override
+    public void preValidate(final ShardingRule shardingRule, final SQLStatementContext<AlterTableStatement> sqlStatementContext, final List<Object> parameters, final ShardingSphereSchema schema) {
+        Collection<String> shardingLogicTableNames = getShardingLogicTableNames(shardingRule, sqlStatementContext);
+        if (shardingLogicTableNames.size() <= 1) {
+            return;
+        }
```
Comment (tristaZero) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-route/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingAlterTableStatementValidator.java:44]: How about `broadcast Table`or `single table`?
Is #10172 a interim solution?

**Review by tristaZero — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-route/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingAlterTableStatementValidator.java (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-route/src/main/java/org/apache/shardingsphere/sharding/route/engine/validator/ddl/impl/ShardingAlterTableStatementValidator.java]: > How about `broadcast Table`or `single table`?
> Is #10172 a interim solution?

@tristaZero Sorry for late reply, I spent some time sorting out the different scenes. According to the table type, there are sharding, broadcast and single table in SS. By combining with Cartesian product, we can get nine combination types.

**Comment by tristaZero** [comment]
@strongduanmu Hi, Your summary is much clear and 
So we can judge whether these SQLs are supported or not by `preValidate` and `postValidate`, right?  `preValidate` is for table type checking, and `postVaidate` is for routed result checking?
This PR does not cover all these cases above, does it? But no rush, we can process step by step.

**Comment by strongduanmu** [comment]
> So we can judge whether these SQLs are supported or not by `preValidate` and `postValidate`, right? `preValidate` is for table type checking, and `postVaidate` is for routed result checking?

@tristaZero Yes, `preValidate` currently verifies the data source of single table, and other scenarios can handle routing well through ShardingTableBroadcastRoutingEngine.

**Comment by strongduanmu** [comment]
> This PR does not cover all these cases above, does it? But no rush, we can process step by step.

@tristaZero At present, all supported table combinations have been covered in the rewrite test case, and other exception scenarios have not been covered.

**Comment by tristaZero** [comment]
Hi ,  so could we merge this PR for more supported cases? But the CI is still failing.

**Comment by strongduanmu** [comment]
> Hi , so could we merge this PR for more supported cases? But the CI is still failing.

@tristaZero OK, I will fix the ci exception.

**Review by tristaZero — APPROVED** [review]

**Pull request merged**